src/joystick/hidapi/SDL_hidapi_sinput.c

```diff
@@ -0,0 +1,745 @@
+
+    return true;
+}
+
+static bool HIDAPI_DriverSInput_RumbleJoystick(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, Uint16 low_frequency_rumble, Uint16 high_frequency_rumble)
+{
+    SINPUT_HAPTIC_S hapticData = { 0 };
+    Uint8 hapticReport[SINPUT_DEVICE_REPORT_COMMAND_SIZE] = { SINPUT_DEVICE_REPORT_ID_OUTPUT_CMDDAT, SINPUT_DEVICE_COMMAND_HAPTIC };
+
+    hapticData.type = 1;
+
+    hapticData.type_1.left.frequency_1 = 85;
+    hapticData.type_1.left.frequency_2 = 170;
+
+    hapticData.type_1.right.frequency_1 = 85;
+    hapticData.type_1.right.frequency_2 = 170;
+
+    // It's really left and right amplitude, not low frequency and high frequency...
+    hapticData.type_1.left.amplitude_1 = low_frequency_rumble;
+    hapticData.type_1.left.amplitude_2 = low_frequency_rumble;
+
+    hapticData.type_1.right.amplitude_1 = high_frequency_rumble;
+    hapticData.type_1.right.amplitude_2 = high_frequency_rumble;
+
+    HapticsType1Pack(&hapticData, &(hapticReport[2]));
+
+    SDL_HIDAPI_SendRumble(device, hapticReport, SINPUT_DEVICE_REPORT_COMMAND_SIZE);
+
+    return true;
+}
+
+static bool HIDAPI_DriverSInput_RumbleJoystickTriggers(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, Uint16 left_rumble, Uint16 right_rumble)
+{
+    return SDL_Unsupported();
+}
+
+static Uint32 HIDAPI_DriverSInput_GetJoystickCapabilities(SDL_HIDAPI_Device *device, SDL_Joystick *joystick)
+{
+    Uint32 caps = 0;
+    caps |= SDL_JOYSTICK_CAP_RUMBLE;
+    caps |= SDL_JOYSTICK_CAP_PLAYER_LED;
+    return caps;
+}
+
+static bool HIDAPI_DriverSInput_SetJoystickLED(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, Uint8 red, Uint8 green, Uint8 blue)
+{
+    return SDL_Unsupported();
+}
+
+static bool HIDAPI_DriverSInput_SendJoystickEffect(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, const void *data, int size)
+{
+    return SDL_Unsupported();
+}
+
+static bool HIDAPI_DriverSInput_SetJoystickSensorsEnabled(SDL_HIDAPI_Device *device, SDL_Joystick *joystick, bool enabled)
+{
+    return true;
+}
+
+static void HIDAPI_DriverSInput_HandleStatePacket(SDL_Joystick *joystick, SDL_DriverSInput_Context *ctx, Uint8 *data, int size)
+{
+    Sint16 axis = 0;
+    Sint16 accel = 0;
+    Sint16 gyro = 0;
+    Uint64 timestamp = SDL_GetTicksNS();
+    float imu_values[3] = { 0 };
+
+    if (ctx->last_state[SINPUT_REPORT_IDX_BUTTONS_0] != data[SINPUT_REPORT_IDX_BUTTONS_0]) {
+        SDL_SendJoystickButton(timestamp, joystick, SDL_GAMEPAD_BUTTON_SOUTH, (data[SINPUT_REPORT_IDX_BUTTONS_0] & 0x01));
+        SDL_SendJoystickButton(timestamp, joystick, SDL_GAMEPAD_BUTTON_EAST,  (data[SINPUT_REPORT_IDX_BUTTONS_0] & 0x02));
+        SDL_SendJoystickButton(timestamp, joystick, SDL_GAMEPAD_BUTTON_WEST,  (data[SINPUT_REPORT_IDX_BUTTONS_0] & 0x04));
+        SDL_SendJoystickButton(timestamp, joystick, SDL_GAMEPAD_BUTTON_NORTH, (data[SINPUT_REPORT_IDX_BUTTONS_0] & 0x08));
+
+        Uint8 hat;
+        switch ((data[SINPUT_REPORT_IDX_BUTTONS_0] & 0xF0) >> 4) {
+        case 0b0001:
```

Comment:
Please use hex values instead of binary values, for consistency with the other drivers.